### PR TITLE
crimson/os/seastore/object_data_handler: cleanup the overwrite path

### PR DESCRIPTION
**Pull request by cyx1231st**
* implement factory and type for extent_to_write_t.
* refactor overwrite_plan_t with explicit size calculations and
  defined operations.
* cleanup the overwrite path accordingly.

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>

The goal is to do cleanups for the possible follow up related changes:
* Fix overwrite with direct read that bypasses the cache layer, also the lazy read: https://github.com/ceph/ceph/pull/45519#discussion_r1065511533 @aisakaki @zhscn.
* Finer-grained read cache or the max-allocation-size support @ljx023 @athanatos.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test api

File: src/crimson/os/seastore/object_data_handler.cc [hunk 1]
```diff
@@ -45,23 +45,51 @@ auto read_pin(
...
+    ZERO,
+    EXISTING,
+  };
+
+  type_t type;
+  laddr_t addr;
   extent_len_t len;
   std::optional<bufferlist> to_write;
-  std::optional<paddr_t> existing_paddr = std::nullopt;
+  std::optional<paddr_t> existing_paddr;
```
Comment (athanatos) [inline src/crimson/os/seastore/object_data_handler.cc:58]: /// non-nullopt if and only if type == EXISTING

File: src/crimson/os/seastore/object_data_handler.cc [hunk 2]
```diff
@@ -136,7 +166,7 @@ ObjectDataHandler::write_ret do_insertions(
     to_write,
     [ctx](auto &region) {
       LOG_PREFIX(object_data_handler.cc::do_insertions);
-      if (region.to_write) {
+      if (region.type == extent_to_write_t::type_t::DATA) {
```
Comment (athanatos) [inline src/crimson/os/seastore/object_data_handler.cc:169]: Might be worth replacing these comparisons with is_data(), is_zero(), is_existing()?

**Review by athanatos — APPROVED** [review]
LGTM otherwise

**Review by xxhdx1985126 — APPROVED** [review]

File: src/crimson/os/seastore/object_data_handler.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/object_data_handler.cc:60]: done

File: src/crimson/os/seastore/object_data_handler.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/object_data_handler.cc]: done

**Comment by cyx1231st** [comment]
make check result:
```
The following tests FAILED:
	 10 - run-tox-mgr-dashboard-lint (Failed)
	 11 - setup-venv-for-mgr-dashboard-check (Failed)
	 13 - run-tox-mgr-dashboard-check (Not Run)
	 16 - run-tox-mgr-dashboard-openapi (Failed)
```

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins retest this please

**Comment by cyx1231st** [comment]
jenkins test make check

**Pull request merged**